File: EtiveMor.OpenImoCiiCalculator/EtiveMor.OpenImoCiiCalculator.Core/Models/MeasurementModels/ShipDdVectorBoundaries.cs
using EtiveMor.OpenImoCiiCalculator.Core.Models.Enums;

namespace EtiveMor.OpenImoCiiCalculator.Core.Models.MeasurementModels
{
    public class ShipDdVectorBoundaries
    {
        /// <summary>
        /// IMO MEPC.354(78) ddvectors for a given year for the specified ship type
        /// </summary>
        /// <param name="shipType">
        /// The type of ship to generate ddvector boundaries for
        /// </param>
        /// <param name="weightClassification">
        /// The weight classification of the ship to generate ddvector boundaries for.
        /// 
        /// If these ddvectors have a min/max weight boundary in MEPC.354(78), this object describes the 
        /// lower and upper bound of that classification. For example, Gas Carriers below
        /// 65000 DWT have different ddvectors to those at or above 65000 DWT
        /// 
        /// If the ddvectors do not have a weight classification, this object will contain
        /// a range between 0 and int.MaxValue
        /// </param>
        /// <param name="capacityUnit">
        /// Indicates the capacity unit these ddvectors are calculated against. For example
        /// GT for passenger/ro-ro ships and DWT for cargo carriers
        /// </param>
        /// <param name="boundaryDdVectors">
        /// The ddvectors for the specified ship type, weight classification and capacity unit in the given year
        /// </param>
        /// <param name="year">
        /// The year these ddvectors apply to. Note that the ddvectors are only valid for the specified
        /// calendar year
        /// </param>
        public ShipDdVectorBoundaries(
            ShipType shipType,
            WeightClassification weightClassification,
            CapacityUnit capacityUnit,
            Dictionary<ImoCiiBoundary, double> boundaryDdVectors, 
            int year)
        {
            ShipType = shipType;
            WeightClassification = weightClassification;
            CapacityUnit = capacityUnit;
            BoundaryDdVectors = boundaryDdVectors;
            Year = year;
        }

        /// <summary>
        /// The year these DDVector boundaries apply to 
        /// </summary>
        public int Year { get; private set; }

        /// <summary>
        /// The shipType these vector boundaries apply to
        /// </summary>
        public ShipType ShipType { get; set; }

        /// <summary>
        /// The weight classification these vector boundaries apply to
        /// </summary>
        public WeightClassification WeightClassification { get; set; }
        
        /// <summary>
        /// The capacity unit these vector boundaries apply to
        /// </summary>
        public CapacityUnit CapacityUnit { get; set; }

        public Dictionary<ImoCiiBoundary, double> BoundaryDdVectors { get; set; }
    }

    public class WeightClassification
    {
        public WeightClassification(int upperLimit, int lowerLimit)
        {
            UpperLimit = upperLimit;
            LowerLimit = lowerLimit;
        }

        public int UpperLimit { get; private set; }
        public int LowerLimit { get; private set; }
    }
}

File: python-impl/open_imo_cii_calculator/models/measurement_models/ship_dd_vector_boundaries.py
from typing import Dict
from open_imo_cii_calculator.models.ship_type import ShipType
from open_imo_cii_calculator.models.capacity_unit import CapacityUnit
from open_imo_cii_calculator.models.imo_cii_boundary import ImoCiiBoundary
from open_imo_cii_calculator.models.measurement_models.weight_classification import WeightClassification

class ShipDdVectorBoundaries:
    """
    IMO MEPC.354(78) ddvectors for a given year for the specified ship type
    """
    
    def __init__(self, 
                 ship_type: ShipType, 
                 weight_classification: WeightClassification,
                 capacity_unit: CapacityUnit,
                 boundary_dd_vectors: Dict[ImoCiiBoundary, float], 
                 year: int):
        """
        Initialize ship ddvector boundaries
        
        Args:
            - ship_type (ShipType): The type of ship to generate ddvector boundaries for
            - weight_classification (WeightClassification): The weight classification of the ship. If these ddvectors have a min/max 
                weight boundary in MEPC.354(78), this describes the lower and upper bound of that classification.
            - capacity_unit (CapacityUnit): Indicates the capacity unit these ddvectors are calculated against
            - boundary_dd_vectors (Dict[ImoCiiBoundary, float]): The ddvectors for the specified ship type, weight classification and capacity unit
            - year (int): The year these ddvectors apply to
        """
        self.year = year
        self.ship_type = ship_type
        self.weight_classification = weight_classification
        self.capacity_unit = capacity_unit
        self.boundary_dd_vectors = boundary_dd_vectors